File: 79-word-search/word-search.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        int m = board.size();
        int n = board[0].size();
        
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (dfs(board, word, i, j, 0)) {
                    return true;
                }
            }
        }
        
        return false;
    }

private:
    bool dfs(vector<vector<char>>& board, string& word, int i, int j, int index) {
        if (index == word.size()) {
            return true;
        }
        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size() || board[i][j] != word[index]) {
            return false;
        }
        
        char temp = board[i][j];
        board[i][j] = '#';  // Mark the cell as visited
        
        bool found = dfs(board, word, i + 1, j, index + 1) || 
                     dfs(board, word, i - 1, j, index + 1) || 
                     dfs(board, word, i, j + 1, index + 1) || 
                     dfs(board, word, i, j - 1, index + 1);
        
        board[i][j] = temp;  // Unmark the cell
        
        return found;
    }
};

File: 79-word-search/word-search.java
class Solution {
    int[] delRow={0,-1,0,1};
    int[] delCol={-1,0,1,0};
    int n,m;
    public boolean backTrack(char[][] board,String word,int row,int col,int ind){
        if(ind==word.length())return true;
        char temp=board[row][col];
        board[row][col]='#';
        for(int i=0;i<4;i++){
            int newRow=row+delRow[i];
            int newCol=col+delCol[i];
            if((newRow>=0&&newRow<n)&&(newCol>=0&&newCol<m)&&(word.charAt(ind)==board[newRow][newCol])){
                if(backTrack(board,word,newRow,newCol,ind+1)){
                    board[row][col]=temp;
                    return true;
                }
            }
        }
        board[row][col]=temp;
        return false;
    }
    public boolean exist(char[][] board, String word) {
        n=board.length;
        m=board[0].length;
        for(int i=0;i<n;i++){
        for(int j=0;j<m;j++){
            if(board[i][j] == word.charAt(0)){
                if(backTrack(board,word,i,j,1)){
                    return true;
                }
            }
        }
        }
        return false;
    }
}